### PR TITLE
Make motion actuators play nicely together

### PR DESCRIPTION
**Pull request by severin-lemaignan**
This is in particular useful for the (soon to come?) new human model, that 'embeds' a keyboard actuator all the time.

### CODE REVIEW

**Comment by adegroote** [comment]
I disagree with the premise (why to embeds a keyboard actuator in human ???), through I'm not completely against this change. However, I'm worried by two issues:
 - 1/ it breaks / changes the semantic of the Waypoint actuator. Previously, it enforces that the robot stays around its waypoint, whatever is its environment. Now, it is definitely reach and forget. The new proposed semantic have some sense, but I would prefer to make it configurable by user.  For input "datastream", we can have basically two semantics:
  - always enforce last command (which is the original semantic of Morse)
  - execute "once" and do nothing before next command (which is what you propose now).
The two semantics has their pro and cons, and it would be definitively be nice to let user choose the good one for this simulation.
 - 2/ it makes behavior of Morse actuators inconsistent (why destination or rotorcraft_waypoint does not have the same behaviour than waypoint ?)

To summarize, I will be more happy if the change is done in a more configurable and homogeneous  way over Morse actuators.

**Comment by severin-lemaignan** [comment]
> I disagree with the premise (why to embeds a keyboard actuator in human ???)

Well, this can be discussed. My current approach, so far, it (1) to remain relatively close to to current human model, to prevent breaking expectations, (2) to make the human avatar as useful as possible 'out of the box': the human avatar simulates a human, ie an agent with a complex behaviour, and in many cases, controlling the human with the keyboard is the best way to get a semi-realistic simulated behaviour.

But even if you forget about the keyboard actuator, I can imagine cases where you may want to have both a `MotionVW` and a `Waypoint` actuator on the same robot for different parts of the simulation.

**Comment by severin-lemaignan** [comment]
> 1/ it breaks / changes the semantic of the Waypoint actuator. Previously, it enforces that the robot stays around its waypoint, whatever is its environment. Now, it is definitely reach and forget. The new proposed semantic have some sense, but I would prefer to make it configurable by user. For input "datastream", we can have basically two semantics:
> 
>     always enforce last command (which is the original semantic of Morse)
>     execute "once" and do nothing before next command (which is what you propose now). The two semantics has their pro and cons, and it would be definitively be nice to let user choose the good one for this simulation.

Ok, I implement that. But I suggest to have the new behaviour that I propose as default, because I would argue this is the expected behaviour: when you send a waypoint to a system, you expect it to go to (are actually 'over') this point. You do not expect the system to go and stick to the target point.

> 
> 2/ it makes behavior of Morse actuators inconsistent (why destination or rotorcraft_waypoint does not have the same behaviour than waypoint ?)

I can also add it for these two actuators.

**Comment by adegroote** [comment]
> But even if you forget about the keyboard actuator, I can imagine cases where you may want to have both a MotionVW and a Waypoint actuator on the same robot for different parts of the simulation.

I don't think it works even after your change (or only by accident), as MotionVW always call apply_speed (with the last received command).

> But I suggest to have the new behaviour that I propose as default, because I would argue this is the expected behaviour: when you send a waypoint to a system, you expect it to go to (are actually 'over') this point. You do not expect the system to go and stick to the target point.

I  assumed the same two years ago, when I care mostly about ground robot, on almost flat ground. Considering now flying or sailing robot, you need to actively control your robot to stay around your point, and you definitively want to do that (particularly for flying robot, where you will just fall if you do nothing).

Considering 2/, the list was not exhaustive, I just want to have consistency over Morse actuator behavior.

**Comment by severin-lemaignan** [comment]
Alright, so this PR is getting larger than initially planned, since it now tries to address the larger issue: can we combine more than one motion actuator on a robot?

Currently:
- Two 'waypoint' actuators (`Waypoint` and `Rotorcraft_waypoint`) have now an option (`RemainAtDestination`) that controls whether the robot attempts to actively remain at the destination waypoint. It is `false` by default for `Waypoint` and `True` for `Rotorcraft_waypoint` (since it is *really* useful for flying robots)
- I've updated `Keyboard` and `Joystick` so that they do not spam the robot's `apply_speed()` method.

The many other remaining motion actuators:
- `Teleport`
- `Orientation`
- `Destination`
- `MotionVW`
- `MotionXYW`
- `MotionVWDiff`
- `RotorcraftAttitude` -> force control
- `RotorcraftVelocity` -> force control
- `StabilizedQuadrotor` -> force control
- `SteerForce` -> force control

The general questions are:
- should we allow at all more than one motion actuator? (if not, 'manual override' from keyboard/joystick could be possibly exceptions)
- if no, we should probably detect when we have more than one actuator and 'complain' about it
- if yes, we need to establish a few rule (who can override what and when)

This discussion should likely happen on morse-dev (if not morse-users), since some users may have use-cases for more than one actuator.

Meanwhile, and assuming no other issues with the current commits, I would love if the PR could be merged without waiting for a complete solution (I need it for other pending PRs :-)

**Comment by adegroote** [comment]
The commits look ok, through the solution seems a little ad-hoc for me. I would like to have also the same change in Destination, as it has globally the same semantic than Waypoint. And you need to add an entry in what_news.rst / CHANGELOG. And we probably can merge in this state.

Concerning the overall discussion, I can think a few cases where we need multiples actuators on the same robot, one considering a simulated drone, with some switch from speed-control to position-control. There is also the case of unit-test where we often use multiples actuators, the tricks is to use activate/deactivate service to select the "right" one. 

About what is written:
  - StabilizedQuadrotor is a position-controller 
  - the rule about one actuator is a bit more complex than that, because you can have multiples moving part on a robot, controlled independently. 

Otherwise, please open a discussion on morse-{whatyouwant}@

**Comment by severin-lemaignan** [comment]
> I would like to have also the same change in Destination, as it has globally the same semantic than Waypoint. 

Done

> And you need to add an entry in what_news.rst / CHANGELOG. 

Done
